Nightly Build Failure 2024-02-08

# Overview

What do you think happened, if anything?

Seems like the integration tests failed because they couldn't find an XBRL SQLite database. Which is weird because the integration tests seem to have passed in other contexts, so how are they being run differently in `gcp_pudl_etl.sh`?

# Next steps

What next steps do we need to do to understand or remediate the issue?

See if we can reproduce this issue locally.

# Verify that everything is fixed!

Once you've applied any necessary fixes, make sure that the nightly build outputs are all in their right places.

```[tasklist]
- [ ] [S3 distribution bucket](https://s3.console.aws.amazon.com/s3/buckets/pudl.catalyst.coop?region=us-west-2&bucketType=general&prefix=nightly/&showversions=false) was updated at the expected time
- [ ] [GCP distribution bucket](https://console.cloud.google.com/storage/browser/pudl.catalyst.coop/nightly;tab=objects?project=catalyst-cooperative-pudl) was updated at the expected time
- [ ] [GCP internal bucket](https://console.cloud.google.com/storage/browser/builds.catalyst.coop) was updated at the expected time
- [ ] [Datasette PUDL version](https://data.catalyst.coop/pudl/core_pudl__codes_datasources) points at the same hash as [nightly](https://github.com/catalyst-cooperative/pudl/tree/nightly)
- [ ] [Zenodo sandbox record](https://sandbox.zenodo.org/doi/10.5072/zenodo.5563) was updated to the record number in the logs (search for `zenodo_data_release.py` and `Draft` in the logs, to see what the new record number should be!)
```

# Relevant logs
[link to build logs from internal distribution bucket](https://console.cloud.google.com/storage/browser/_details/builds.catalyst.coop/2024-02-08-0603-599b509da-main/2024-02-08-0603-599b509da-main-pudl-etl.log)

```
=================================== FAILURES ===================================
_______________ TestFerc1ExtractDebugFunctions.test_extract_xbrl _______________
[gw0] linux -- Python 3.11.7 /home/mambauser/env/bin/python3.11

self = <test.integration.etl_test.TestFerc1ExtractDebugFunctions object at 0x7a13e9964250>
ferc1_engine_dbf = Engine(sqlite:////home/mambauser/pudl_work/output/ferc1_dbf.sqlite)

    def test_extract_xbrl(self, ferc1_engine_dbf):
        """Test extract_xbrl."""
        years = [2021]  # add desired years here
        configured_dataset_settings = {"ferc1": {"years": years}}
    
        dataset_init_context = build_init_resource_context(
            config=configured_dataset_settings
        )
        configured_dataset_settings = pudl.resources.dataset_settings(
            dataset_init_context
        )
    
>       ferc1_xbrl_raw_dfs = pudl.extract.ferc1.extract_xbrl(
            configured_dataset_settings
        )

test/integration/etl_test.py:206: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/pudl/extract/ferc1.py:558: in extract_xbrl
    io_manager = ferc1_xbrl_sqlite_io_manager(io_manager_init_context)
../env/lib/python3.11/site-packages/dagster/_core/definitions/resource_definition.py:269: in __call__
    return resource_invocation_result(
../env/lib/python3.11/site-packages/dagster/_core/definitions/resource_invocation.py:32: in resource_invocation_result
    resource_fn(_init_context) if has_at_least_one_parameter(resource_fn) else resource_fn()  # type: ignore  # (strict type guard)
src/pudl/io_managers.py:836: in ferc1_xbrl_sqlite_io_manager
    return FercXBRLSQLiteIOManager(
src/pudl/io_managers.py:570: in __init__
    super().__init__(base_dir, db_name, md, timeout)
src/pudl/io_managers.py:157: in __init__
    self.engine = self._setup_database(timeout=timeout)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pudl.io_managers.FercXBRLSQLiteIOManager object at 0x7a13e7590390>
timeout = 1000.0

    def _setup_database(self, timeout: float = 1_000.0) -> sa.Engine:
        """Create database engine and read the metadata.
    
        Args:
            timeout: How many seconds the connection should wait before raising an
                exception, if the database is locked by another connection.  If another
                connection opens a transaction to modify the database, it will be locked
                until that transaction is committed.
    
        Returns:
            engine: SQL Alchemy engine that connects to a database in the base_dir.
        """
        # If the sqlite directory doesn't exist, create it.
        db_path = self.base_dir / f"{self.db_name}.sqlite"
        if not db_path.exists():
>           raise ValueError(
                f"No DB found at {db_path}. Run the job that creates the "
                f"{self.db_name} database."
            )
E           ValueError: No DB found at /home/mambauser/pudl_work/output/ferc1_xbrl.sqlite. Run the job that creates the ferc1_xbrl database.

src/pudl/io_managers.py:587: ValueError
```

--- a/test/conftest.py
+++ b/test/conftest.py
@@ -321,11 +321,18 @@ def pudl_engine(pudl_io_manager: PudlMixedFormatIOManager) -> sa.Engine:
 def configure_paths_for_tests(tmp_path_factory, request):
     """Configures PudlPaths for tests.
 
-    Typically PUDL_INPUT and PUDL_OUTPUT will be read from the environment.
-    If we are running in GitHub Actions and they are NOT set, we'll use temp dirs.
-    If we are NOT running in GitHub Actions (e.g. we're running locally) then we always
-    want to use a temporary output directory, so we don't overwrite a user's existing
-    databases.
+    Default behavior:
+
+    PUDL_INPUT is read from the environment.
+    PUDL_OUTPUT is set to a tmp path, to avoid clobbering existing databases.
+
+    Set ``--tmp-data`` to force PUDL_INPUT to a temporary directory, causing
+    re-downloads of all raw inputs.
+
+    Set ``--live-dbs`` to force PUDL_OUTPUT to *NOT* be a temporary directory
+    and instead inherit from environment.
+
+    ``--live--dbs`` flag is ignored in unit tests, see pudl/test/unit/conftest.py.
     """
     # Just in case we need this later...
     pudl_tmpdir = tmp_path_factory.mktemp("pudl")

--- a/test/unit/conftest.py
+++ b/test/unit/conftest.py
@@ -1,0 +1,53 @@
+import logging
+from pathlib import Path
+
+import pydantic
+import pytest
+
+from pudl.workspace.setup import PudlPaths
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture(scope="session", autouse=True)
+def configure_paths_for_tests(tmp_path_factory, request):
+    """Configures PudlPaths for tests.
+
+    Default behavior:
+
+    PUDL_INPUT is read from the environment.
+    PUDL_OUTPUT is set to a tmp path, to avoid clobbering existing databases.
+
+    Set ``--tmp-data`` to force PUDL_INPUT to a temporary directory, causing
+    re-downloads of all raw inputs.
+
+    Ignores the ``--live-dbs`` flag; always forces PUDL_OUTPUT to a temp dir so
+    unit test can never mess with the outputs.
+
+    See pudl/test/conftest.py for the non-unit test counterpart.
+    """
+    pudl_tmpdir = tmp_path_factory.mktemp("pudl")
+
+    # We only use a temporary input directory when explicitly requested.
+    # This will force a re-download of raw inputs from Zenodo or the GCS cache.
+    if request.config.getoption("--tmp-data"):
+        in_tmp = pudl_tmpdir / "input"
+        in_tmp.mkdir()
+        PudlPaths.set_path_overrides(
+            input_dir=str(Path(in_tmp).resolve()),
+        )
+        logger.info(f"Using temporary PUDL_INPUT: {in_tmp}")
+
+    out_tmp = pudl_tmpdir / "output"
+    out_tmp.mkdir()
+    PudlPaths.set_path_overrides(
+        output_dir=str(Path(out_tmp).resolve()),
+    )
+    logger.info(f"Using temporary PUDL_OUTPUT: {out_tmp}")
+
+    try:
+        return PudlPaths()
+    except pydantic.ValidationError as err:
+        pytest.exit(
+            f"Set PUDL_INPUT, PUDL_OUTPUT env variables, or use --tmp-path, --live-dbs flags. Error: {err}."
+        )

--- a/test/unit/extract/xbrl_test.py
+++ b/test/unit/extract/xbrl_test.py
@@ -1,6 +1,5 @@
 """Tests for xbrl extraction module."""
 
-import os
 
 import pytest
 from dagster import ResourceDefinition, build_op_context
@@ -110,8 +109,6 @@ def test_xbrl2sqlite(settings, forms, mocker, tmp_path):
         if op.tags.get("data_format") == "xbrl"
     ]
 
-    # always use tmp path here so that we don't clobber the live DB when --live-dbs is passed
-    mocker.patch.dict(os.environ, {"PUDL_OUTPUT": tmp_path})
     ferc_to_sqlite.execute_in_process(
         op_selection=op_selection,
         resources={
@@ -140,9 +137,6 @@ def test_xbrl2sqlite(settings, forms, mocker, tmp_path):
 
 
 def test_xbrl2sqlite_db_exists_no_clobber(mocker, tmp_path):
-    # always use tmp path here so that we don't clobber the live DB when --live-dbs is passed
-    mocker.patch.dict(os.environ, {"PUDL_OUTPUT": tmp_path})
-
     convert_form_mock = mocker.MagicMock()
     mocker.patch("pudl.extract.xbrl.convert_form", new=convert_form_mock)
 
@@ -174,9 +168,6 @@ def test_xbrl2sqlite_db_exists_no_clobber(mocker, tmp_path):
 
 
 def test_xbrl2sqlite_db_exists_yes_clobber(mocker, tmp_path):
-    # always use tmp path here so that we don't clobber the live DB when --live-dbs is passed
-    mocker.patch.dict(os.environ, {"PUDL_OUTPUT": tmp_path})
-
     convert_form_mock = mocker.MagicMock()
     mocker.patch("pudl.extract.xbrl.convert_form", new=convert_form_mock)
 

--- a/test/unit/extract/xbrl_test.py
+++ b/test/unit/extract/xbrl_test.py
@@ -1,5 +1,7 @@
 """Tests for xbrl extraction module."""
 
+import os
+
 import pytest
 from dagster import ResourceDefinition, build_op_context
 
@@ -109,6 +111,7 @@ def test_xbrl2sqlite(settings, forms, mocker, tmp_path):
     ]
 
     # always use tmp path here so that we don't clobber the live DB when --live-dbs is passed
+    mocker.patch.dict(os.environ, {"PUDL_OUTPUT": tmp_path})
     ferc_to_sqlite.execute_in_process(
         op_selection=op_selection,
         resources={
@@ -121,9 +124,6 @@ def test_xbrl2sqlite(settings, forms, mocker, tmp_path):
             ),
         },
     )
-
-    # TODO(rousik): do we need to use this, or can we simply set PUDL_OUTPUT env
-    # variable to some random path?
 
     assert convert_form_mock.call_count == len(forms)
 
@@ -139,9 +139,9 @@ def test_xbrl2sqlite(settings, forms, mocker, tmp_path):
         )
 
 
-def test_xbrl2sqlite_db_exists_no_clobber(mocker, live_dbs):
-    if live_dbs:
-        return
+def test_xbrl2sqlite_db_exists_no_clobber(mocker, tmp_path):
+    # always use tmp path here so that we don't clobber the live DB when --live-dbs is passed
+    mocker.patch.dict(os.environ, {"PUDL_OUTPUT": tmp_path})
 
     convert_form_mock = mocker.MagicMock()
     mocker.patch("pudl.extract.xbrl.convert_form", new=convert_form_mock)
@@ -173,9 +173,9 @@ def test_xbrl2sqlite_db_exists_no_clobber(mocker, live_dbs):
     assert ferc1_sqlite_path.exists()
 
 
-def test_xbrl2sqlite_db_exists_yes_clobber(mocker, live_dbs):
-    if live_dbs:
-        return
+def test_xbrl2sqlite_db_exists_yes_clobber(mocker, tmp_path):
+    # always use tmp path here so that we don't clobber the live DB when --live-dbs is passed
+    mocker.patch.dict(os.environ, {"PUDL_OUTPUT": tmp_path})
 
     convert_form_mock = mocker.MagicMock()
     mocker.patch("pudl.extract.xbrl.convert_form", new=convert_form_mock)
@@ -184,7 +184,6 @@ def test_xbrl2sqlite_db_exists_yes_clobber(mocker, live_dbs):
     mock_datastore = mocker.MagicMock()
     mocker.patch("pudl.extract.xbrl.FercXbrlDatastore", return_value=mock_datastore)
 
-    # always use tmp path here so that we don't clobber the live DB when --live-dbs is passed
     ferc1_sqlite_path = PudlPaths().output_dir / "ferc1_xbrl.sqlite"
     ferc1_sqlite_path.touch()
     settings = FercToSqliteSettings(

